process/block/baseProcess.go

```diff
@@ -880,43 +632,52 @@ func (bp *baseProcessor) requestHeaderByShardAndNonce(targetShardID uint32, nonc
 	}
 }
 
-func (bp *baseProcessor) isShardStuck(shardId uint32) bool {
-	header := bp.lastHdrForShard(shardId)
-	if check.IfNil(header) {
-		return false
-	}
-
-	isShardStuck := bp.rounder.Index()-int64(header.GetRound()) > process.MaxRoundsWithoutCommittedBlock
-	return isShardStuck
-}
-
 func (bp *baseProcessor) cleanupPools(
+	headerHandler data.HeaderHandler,
 	headersNoncesPool dataRetriever.Uint64SyncMapCacher,
 	headersPool storage.Cacher,
 	notarizedHeadersPool storage.Cacher,
 ) {
+	noncesToFinal := bp.getNoncesToFinal(headerHandler)
+
 	bp.removeHeadersBehindNonceFromPools(
 		true,
 		headersPool,
 		headersNoncesPool,
 		bp.shardCoordinator.SelfId(),
 		bp.forkDetector.GetHighestFinalBlockNonce())
 
-	for shardId := range bp.notarizedHdrs {
-		lastNotarizedHdr := bp.lastNotarizedHdrForShard(shardId)
-		if check.IfNil(lastNotarizedHdr) {
+	for shardID := uint32(0); shardID < bp.shardCoordinator.NumberOfShards(); shardID++ {
+		if bp.shardCoordinator.SelfId() == shardID {
 			continue
 		}
 
-		bp.removeHeadersBehindNonceFromPools(
-			false,
-			notarizedHeadersPool,
-			headersNoncesPool,
-			shardId,
-			lastNotarizedHdr.GetNonce())
+		bp.cleanupPoolsForShard(shardID, headersNoncesPool, notarizedHeadersPool, noncesToFinal)
```

Comment:
Yes, because when you want to cleanup cross shard notarized headers you should do this tacking in account how many roll backs you still can do in your shard, just because you might need this old cross shard notarized information. So, if you are a node in one shard and you just committed block with nonce 100 and the final is 90, you can still roll back 10 blocks, so you need at least the last 10 meta blocks in the tracker list (each commit in shard means another append in tracker list with a new or the same last notarized meta header). So with this variable noncesToFinal I will call bp.blockTracker.GetCrossNotarizedHeader(shardID, noncesToFinal) and I will receive the meta header on this position counted from the last to first. So now I will take its nonce and I can safely remove all the records behind it.

Reply:
The Final Header should be kept by blockTracker - so the cleanup should be called by him, no need for baseProcess to give this parameter. Cleanup would be called after every commit, checking the final block which is irreversible from blocktracker.

Reply:
Block tracker would be always ahead of the shard/meta processing nonce. If it would take the cleaning decision based on itself state, it would delete things that you would need after some time. So it would need an input from the processor to know where the processing is and to cleanup only behind the processed/committed/finalized nonce. For example, if you are a shard node, and you are at nonce 100 and your final nonce based on what you have already been processed/committed (meta blocks included in the shard headers) is only 90 you want to delete headers from block tracker for your shard only behind nonce 90. If block tracker would take by itself this decision as it has the latest meta in which it is notarized let's say nonce 98 from your shard, it would delete everything behind nonce 98, and it couldn't serve you with this information gap between 91-97 which you would need very soon.